test: fix failure in feature_nulldummy.py on single-core machines

theStack opened the pull request:
On single-core machines, executing the test `feature_nulldummy.py` results in the following assertion error:

```
    ...
    2021-08-18T15:37:58.805000Z TestFramework (INFO): Test 4: Non-NULLDUMMY base multisig transaction is invalid after activation
    2021-08-18T15:37:58.814000Z TestFramework (ERROR): Assertion failed
    Traceback (most recent call last):
      File "[...]/test/functional/test_framework/test_framework.py", line 131, in main
        self.run_test()
      File "[...]/test/functional/feature_nulldummy.py", line 107, in run_test
        self.block_submit(self.nodes[0], [test4tx], accept=False)
      File "[...]/test/functional/feature_nulldummy.py", line 134, in block_submit
        assert_equal(None if accept else 'block-validation-failed', node.submitblock(block.serialize().hex()))
      File "[...]/test/functional/test_framework/util.py", line 49, in assert_equal
        raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
    AssertionError: not(block-validation-failed == non-mandatory-script-verify-flag (Dummy CHECKMULTISIG argument must be zero))
    2021-08-18T15:37:58.866000Z TestFramework (INFO): Stopping nodes
    ...
```

There are hardly any single-core machines around anymore, but the behaviour can be reproduced on a multi-core machine by patching the function `GetNumCores()` to return 1 on the master branch and running `feature_nulldummy.py`:

```diff
diff --git a/src/util/system.cpp b/src/util/system.cpp
index 30d410381..149b512fc 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -1338,7 +1338,7 @@ bool SetupNetworking()

 int GetNumCores()
 {
-    return std::thread::hardware_concurrency();
+    return 1;
 }
```
As solution, parallel script verification is disabled (`-par=1`) and the exact reject reason is checked, which also increases the precision of the test (the possibility that the block is rejected because of another unintended reason is ruled out). See also related PR #22711 which applies the same approach for the p2p segwit test. The PR also includes a refactoring commit which changes the calls to `self.block_submit()` to use named arguments and removes the default value for parameter `accept` (i.e. explicitely passing `accept=...` is mandatory), with the aim to increase the test readability.

theStack committed:
test: refactor: use named args for block_submit in feature_nulldummy.py

theStack committed:
test: fix failure in feature_nulldummy.py on single-core machines

On single-core machines, executing the test feature_nulldummy.py results in
the following assertion error:

...
2021-08-18T15:37:58.805000Z TestFramework (INFO): Test 4: Non-NULLDUMMY base multisig transaction is invalid after activation
2021-08-18T15:37:58.814000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "[...]/test/functional/test_framework/test_framework.py", line 131, in main
    self.run_test()
  File "[...]/test/functional/feature_nulldummy.py", line 107, in run_test
    self.block_submit(self.nodes[0], [test4tx], accept=False)
  File "[...]/test/functional/feature_nulldummy.py", line 134, in block_submit
    assert_equal(None if accept else 'block-validation-failed', node.submitblock(block.serialize().hex()))
  File "[...]/test/functional/test_framework/util.py", line 49, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(block-validation-failed == non-mandatory-script-verify-flag (Dummy CHECKMULTISIG argument must be zero))
2021-08-18T15:37:58.866000Z TestFramework (INFO): Stopping nodes
...

The behaviour can be reproduced on a multi-core machine by simply changing the
function GetNumCores() (in src/util/system.cpp) to return 1:

int GetNumCores()
{
    return 1;
}

josibake commented on test/functional/feature_nulldummy.py:
```diff
@@ -24,7 +24,10 @@
 from test_framework.script import CScript
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_equal, assert_raises_rpc_error
+from test_framework.util import (
+    assert_equal,
+    assert_raises_rpc_error,
+)
```
nit: style only change, not a big deal since this is a very small pr

josibake commented on test/functional/feature_nulldummy.py:
```diff
@@ -86,36 +89,36 @@ def run_test(self):
 
         self.log.info("Test 4: Non-NULLDUMMY base multisig transaction is invalid after activation")
         test4tx = create_transaction(self.nodes[0], test2tx.hash, self.address, amount=46)
         test6txs = [CTransaction(test4tx)]
         trueDummy(test4tx)
         assert_raises_rpc_error(-26, NULLDUMMY_ERROR, self.nodes[0].sendrawtransaction, test4tx.serialize_with_witness().hex(), 0)
-        self.block_submit(self.nodes[0], [test4tx])
+        self.block_submit(self.nodes[0], [test4tx], accept=False)
```
can you rename this commit to `refactor: ...`? i stared at this line for longer than id like to admit, trying to understand why you were adding a new argument and how it was changing the test behavior :sweat_smile:

theStack replied:
Good point, I forgot this in the first commit. Sorry for creating confusion! Done.

josibake reviewed (commented):
Concept ACK

all for improving test precision by using specific error messages. i was unable, however, to reproduce the issue. i re-compiled with the change suggested in the description (`return 1`) and ran the test with and without `-par=1`. it passed both times, but maybe im not doing something right?

id also suggest renaming the first commit as a `refactor` and updating the description to note you are fixing the test and doing a small refactor. without that context, its a little confusing as to why the arguments are changing

theStack commented:
@josibake: Thanks a lot for your review! I agree that the description was a bit sloppy/confusing w.r.t. how to reproduce the issue and not mentioning the refactoring commit. I updated the PR description and force-pushed with the following changes:
* rebased on master
* renamed first commit subject to include "refactor:" (as suggested https://github.com/bitcoin/bitcoin/pull/22738#pullrequestreview-736244239)

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/22738/commits/7720d4f650015272dc7109238230520f71858c6c

compiled with the PR suggestion and ensured i can reproduce the issue. re-compiled normally and verified test passes. great work on using a more specific error and improving the readability!

Saviour1001 commented:
Tested ACK <code>[7720d4f](https://github.com/bitcoin/bitcoin/commit/7720d4f650015272dc7109238230520f71858c6c)</code>
